### ISSUE TITLE
No way to use custom Sampler

### ISSUE THREAD

**Issue by dnouri** [dnouri/skorch#99]
There's currently no way to use a custom [Sampler](https://github.com/pytorch/pytorch/blob/master/torch/utils/data/sampler.py).

**Comment by benjamin-work** [comment]
It is possible to pass a sampler to `DataLoader`:

```
net = NeuralNet(..., iterator_train__sampler=mysampler)
```

It is not easily possible for this sampler to reference the `Dataset`.

**Comment by ottonemo** [comment]
This is a problem when using `torch.utils.data.distributed.DistributedSampler` or using your own sampler that depends on the dataset.

Theoretically `SequentialSampler` and `RandomSampler` are problematic as well but these can be set implicitly using the `shuffle` parameter of the `DataLoader`.

**Comment by benjamin-work** [comment]
It is now possible to pass an instantiated dataset. Therefore, this should work:

```
dataset = MyDataset(X, y)
sampler = torch.utils.data.distributed.DistributedSampler(dataset=dataset )
net = NeuralNet(..., dataset=dataset , iterator_train__sampler=sampler)
```

**Comment by ottonemo** [comment]
It is possible to change the instantiation of the `DataLoader`, enabling you to provide samplers that depend on the dataset by doing something like this:

     Net(iterator_train=lambda ds: DataLoader(ds, sampler=SequentialSampler(ds))

**Comment by ottonemo** [comment]
@dnouri Does this resolve your issue?

**Comment by dnouri** [comment]
Cool, thanks.  This would solve my issue.
